Support scrolling while dragging

eugeniodepalo opened the pull request:
This PR adds support for scrolling the list container (or any other parent) while the mouse is dragging an item. It does so by remembering what was the offset of the element relative to the document when the drag starts, and by calculating the new offset on every mouse move.

## Before

![p](https://cloud.githubusercontent.com/assets/151741/8778587/b0361dd6-2ef6-11e5-9d81-3c4e6f3ef2c7.gif)

## After

![c](https://cloud.githubusercontent.com/assets/151741/8778591/b564343c-2ef6-11e5-973b-65a1464dba72.gif)

eugeniodepalo commented on tests/acceptance/smoke-test.js:
```diff
@@ -67,6 +70,36 @@ test('reordering with mouse events', function(assert) {
+      return { y: itemHeight() + 1 };
+    },
+    {
+      dragend: function() {
```
At first I tried adding these sort of callbacks to the `reorder` helper, but that didn't really work because this test is has to rely on the offset to make sense, and `reorder` works by targeting the elements directly.

jgwhite commented:
Awesome. Will review tonight.

jgwhite commented on addon/mixins/sortable-item.js:
```diff
@@ -308,14 +308,18 @@ export default Mixin.create({
       return event => {
         let dx = getX(event) - dragOrigin;
-        let x = elementOrigin + dx;
+        let elementX = this.get('x') - this.element.offsetLeft;
```
Is `this.element.offsetLeft` different to `$(this.element).offset().left`?

eugeniodepalo replied:
Yes, `offsetLeft` returns the distance relative to the offset parent, and will also exclude transforms, while `$(element).offset().left` returns the distance relative to the document.
So by subtracting `this.element.offsetLeft` from `this.get('x')` you basically get the transform offset. I agree that this is kind of a hack and hard to follow - I think I should instead store the original position somewhere every time we set `x`/`y`.

I'm doing this because in order to find the difference in scroll position, the reference point must be the same. If I just do `scrollOrigin - $(this.element).offset().left`, this is not going to yield the scroll offset, because the `left` property will also include the transform previously set to the element.

I'm really not sure if there are better ways to do this - would gladly appreciate any input you have :)

jgwhite replied:
Gotcha. I don’t know if there’s a better approach either. Maybe let’s just add some comments to illuminate?

jgwhite commented on addon/mixins/sortable-item.js:
```diff
@@ -308,14 +308,18 @@ export default Mixin.create({
-        let x = elementOrigin + dx;
+        let elementX = this.get('x') - this.element.offsetLeft;
+        let scrollX = $(this.element).offset().left - elementX;
+        let x = elementOrigin + dx + (scrollOrigin - scrollX);
```
Could you walk me through what’s happening on lines 320–322? I can’t quite picture it :bowtie:

jgwhite replied:
The `x` computed property uses scroll offset (line 120). Can the usage be unified? Should it be applied to the `y` computed property also?

eugeniodepalo replied:
Here we basically account for the scroll to determine the real `dx`, instead of just relying on the distance from the mouse starting point (which will not change if you scroll an element that is parent of the list). So the actual `dx` is the mouse offset + the difference in scroll from when you started dragging.

Once again, not entirely sure this is the best approach!

jgwhite commented on addon/mixins/sortable-item.js:
```diff
@@ -431,17 +438,16 @@ export default Mixin.create({
-    });
+    } else {
+      return Promise.resolve();
+    }
```
Remind me why we need this conditional… was it to make integration testing easier?

eugeniodepalo replied:
Yes, without this the integration "unit test" wasn't working, because that kind of test doesn't actually wait for promises to solve. I think I'm free to remove this now that we switched to a proper acceptance test.

jgwhite commented on tests/helpers/drag.js:
```diff
@@ -0,0 +1,82 @@
+  } else if (mode === 'touch') {
+    start = 'mousedown';
+    move = 'mousemove';
+    end = 'mouseup';
```
I just realised that I messed up my reorder helper and never used the touchevent names!

jgwhite replied:
I have fixed this on master in reorder. These should be `touchstart`, `touchmove`, `touchend`.

eugeniodepalo replied:
Ahah I didn't review it while copying it - thanks will rebase again!

jgwhite commented on tests/helpers/drag.js:
```diff
@@ -0,0 +1,82 @@
+      drag(
+        'mouse',
+        '.some-list li[data-item=uno]',
+        { y: 50 }
```
The third argument should be a function, right?

eugeniodepalo replied:
Woops - good catch :)

jgwhite commented on tests/helpers/drag.js:
```diff
@@ -0,0 +1,82 @@
+      pageX: itemOffset.left,
+      pageY: itemOffset.top
+    }).then(() => {
+      if (callbacks.dragstart != null) {
```
No need to `!= null`.

jgwhite commented on tests/helpers/drag.js:
```diff
@@ -0,0 +1,82 @@
+      pageX: itemOffset.left,
+      pageY: itemOffset.top
+    }).then(() => {
+      if (callbacks.dragmove != null) {
```
No need to `!= null`.

jgwhite commented on tests/helpers/drag.js:
```diff
@@ -0,0 +1,82 @@
+      pageX: itemOffset.left + offset.x,
+      pageY: itemOffset.top + offset.y
+    }).then(() => {
+      if (callbacks.dragend != null) {
```
No need to `!= null`.

jgwhite commented:
@eugeniodepalo the `drag` test helper is a really good idea. I wonder if we can implement `reorder` in terms of `drag`?

eugeniodepalo commented:
> I wonder if we can implement reorder in terms of drag?

Totally agree - I'll try to refactor `reorder` as a special case of drag.

jgwhite commented:
I’m happy we’re taking time over this PR. Lots of solid improvements are coming from it.

eugeniodepalo commented:
@jgwhite I refactored the reorder helper to use drag, how does it look?

I've also simplified the logic that calculates the scroll offset by just using the parent element so that the element transform doesn't mess with our numbers. Do you feel it's easier to understand now?

jgwhite commented:
Looking very cool indeed. I’m going to play around with this a little on my machine to make sure I understand all the moving parts but I think we’re pretty much ready.

mehulkar commented:
Is this good to merge?

jgwhite commented:
@mehulkar not yet. I’m going to do a thorough review at tonight’s Ember London project night. Want to be sure I really understand the code before merging.

jgwhite commented:
Reviewed and ready.

mehulkar commented:
do i have to do anything special to make use of this? I was using `v1.5.3` and just updated to `1.6.2`, but the page doesn't scroll when I try to drag a list item further below the page. Not sure what to expect.

jgwhite commented:
@mehulkar this patch does not enable automatic scrolling, but rather makes ember-sortable aware of scroll offsets. An important piece of prerequisite work to get smart scrolling implemented.
